From 5f8cfa680610c8a38a56283e42ff3d04b4c6391b Mon Sep 17 00:00:00 2001
From: Sugar Zhang <sugar.zhang@rock-chips.com>
Date: Tue, 31 Mar 2020 15:36:49 +0800
Subject: [PATCH] ARM: Make TEXT_OFFSET more flexible

arch/arm/boot/compressed/head.S:209: Error: invalid constant (608000) after fixup

Signed-off-by: Sugar Zhang <sugar.zhang@rock-chips.com>
Change-Id: I1efb510817d02732e76976058e5af97586ea0147
---
 arch/arm/boot/compressed/head.S | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/arch/arm/boot/compressed/head.S b/arch/arm/boot/compressed/head.S
index a2d1e130e6d2..649d57fc795f 100644
--- a/arch/arm/boot/compressed/head.S
+++ b/arch/arm/boot/compressed/head.S
@@ -206,7 +206,8 @@ not_angel:
 		mov	r4, pc
 		and	r4, r4, #0xf8000000
 		/* Determine final kernel image address. */
-		add	r4, r4, #TEXT_OFFSET
+		add	r4, r4, #(TEXT_OFFSET & 0xffff0000)
+		add	r4, r4, #(TEXT_OFFSET & 0x0000ffff)
 #else
 		ldr	r4, =zreladdr
 #endif
@@ -337,7 +338,8 @@ restart:	adr	r0, LC0
 		 * of RAM and hope for the best.
 		 */
 		cmp	r0, #1
-		sub	r0, r4, #TEXT_OFFSET
+		sub	r0, r4, #(TEXT_OFFSET & 0xffff0000)
+		sub	r0, r0, #(TEXT_OFFSET & 0x0000ffff)
 		bic	r0, r0, #1
 		add	r0, r0, #0x100
 		mov	r1, r6
-- 
2.35.3

